pkg/cloudprovider/providers/openstack/openstack_volumes.go

```diff
@@ -215,40 +215,50 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
 	if err != nil {
 		return "", err
 	}
+
 	if volume.Status != VolumeAvailableStatus {
 		errmsg := fmt.Sprintf("volume %s status is %s, not %s, can not be attached to instance %s.", volume.Name, volume.Status, VolumeAvailableStatus, instanceID)
 		glog.Errorf(errmsg)
 		return "", errors.New(errmsg)
 	}
-	cClient, err := os.NewComputeV2()
-	if err != nil {
-		return "", err
-	}
 
 	if volume.AttachedServerId != "" {
 		if instanceID == volume.AttachedServerId {
 			glog.V(4).Infof("Disk %s is already attached to instance %s", volumeID, instanceID)
 			return volume.ID, nil
 		}
-		glog.V(2).Infof("Disk %s is attached to a different instance (%s), detaching", volumeID, volume.AttachedServerId)
-		err = os.DetachDisk(volume.AttachedServerId, volumeID)
-		if err != nil {
-			return "", err
-		}
+		errmsg := fmt.Sprintf("Disk %s is already attached to a different instance (%s)", volumeID, volume.AttachedServerId)
+		glog.V(2).Infof(errmsg)
+		return "", errors.New(errmsg)
 	}
 
-	startTime := time.Now()
-	// add read only flag here if possible spothanis
-	_, err = volumeattach.Create(cClient, instanceID, &volumeattach.CreateOpts{
-		VolumeID: volume.ID,
-	}).Extract()
-	timeTaken := time.Since(startTime).Seconds()
-	recordOpenstackOperationMetric("attach_disk", timeTaken, err)
+	alreadyAttached, err := os.DiskIsAttached(instanceID, volumeID)
```

Comment:
Sorry for late reply. It is too late in china 5 hours ago.  Yeah, jing sent her pr earlier.
I like the change to new client when needed. And we need test cases and e2es to test this. Would you like to help?